Src/Tools/DiscoveryDocPatcher/Patcher.cs

```diff
@@ -0,0 +1,96 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+using Newtonsoft.Json.Linq;
+using System;
+using System.IO;
+
+namespace DiscoveryDocPatcher
+{
+    class Patcher
+    {
+        public static Patcher Load(string path) =>
+            new Patcher(path, JObject.Parse(File.ReadAllText(path)));
+
+        private readonly string _path;
+        private readonly JObject _json;
+        private bool changed = false;
+
+        private Patcher(string path, JObject json)
+        {
+            _path = path;
+            _json = json;
+        }
+
+        public void SaveWithBackup()
+        {
+            if (changed)
+            {
+                File.Move(_path, _path + ".original");
+                string json = _json.ToString();
+                File.WriteAllText(_path, json);
+            }
+        }
+
+        public void Remove(string jPath, JToken oldToken, JToken replacedToken)
+        {
+            // replacedToken is used if this remove operation has already been carried
+            // out on this jPath. It contains the token that is expected to be at the
+            // jPath *after* the remove operation.
+            var token = _json.SelectToken(jPath);
+            if (!JToken.DeepEquals(token, oldToken))
+            {
+                if (JToken.DeepEquals(token, replacedToken))
+                {
+                    // Already replaced, so do nothing.
+                    return;
+                }
+                throw new InvalidOperationException("Content wrong");
+            }
+            changed = true;
+            token.Remove();
+        }
+
+        public void Remove(string jPath, string oldContent, string replacedContent)
+        {
+            JToken oldToken = JToken.Parse(oldContent.Replace('\'', '"'));
+            JToken replacedToken = JToken.Parse(replacedContent.Replace('\'', '"'));
+            Remove(jPath, oldToken, replacedToken);
+        }
+
+        public void Replace(string jPath, JToken oldToken, JToken newToken)
+        {
+            var token = _json.SelectToken(jPath);
+            if (!JToken.DeepEquals(token, oldToken))
+            {
+                if (JToken.DeepEquals(token, newToken))
+                {
+                    // Already replaced, so do nothing.
+                    return;
+                }
+                throw new InvalidOperationException("Content wrong");
+            }
+            changed = true;
+            token.Replace(newToken);
+        }
+
+        public void Replace(string jPath, string oldContent, string newContent)
+        {
+            JToken oldToken = JToken.Parse(oldContent.Replace('\'', '"'));
+            JToken newToken = JToken.Parse(newContent.Replace('\'', '"'));
```

Comment:
Having single quotes stand in for double quotes in the JSON strings being passed in so they don't have to all be escaped in the C# string literals seems reasonable (assuming the content doesn't otherwise contain single quotes, which is currently true), but it might be worth an explanatory comment.